src/bit_machine/mod.rs

```diff
@@ -366,66 +367,79 @@ impl BitMachine {
+
+            // Write bits from active read frame into input frame
+            let buffer_end = buffer.as_mut_ptr().add(buffer_width);
+            let mut write_frame = CFrameItem::new_write(bit_width, buffer_end);
+            for _ in 0..bit_width {
+                let bit = mac.read_bit();
+                c_writeBit(&mut write_frame, bit);
+            }
+            mac.back(bit_width);
 
-        // Sanity Check: This should never really fail, but still good to do
-        if !simplicity_sys::c_jets::sanity_checks() {
-            return Err(JetFailed);
+            // Return input frame as read frame
+            let buffer_ptr = buffer.as_mut_ptr();
+            CFrameItem::new_read(bit_width, buffer_ptr)
         }
-        let src_ty_bit_width = jet.source_ty().to_bit_width();
-        let target_ty_bit_width = jet.target_ty().to_bit_width();
-
-        let a_frame_size = uword_width(src_ty_bit_width);
-        let b_frame_size = uword_width(target_ty_bit_width);
-        // a_frame_size + b_frame_size must be non-zero unless it is a unit to unit jet
-        if a_frame_size == 0 && b_frame_size == 0 {
-            return Ok(());
+
+        /// Return a write frame that is equivalent to the active write frame.
+        unsafe fn get_output_frame(bit_width: usize, buffer: &mut Vec<UWORD>) -> CFrameItem {
+            let buffer_width = uword_width(bit_width);
+            buffer.clear();
+            buffer.resize(buffer_width, 0);
+
+            // Return output frame as write frame
+            let buffer_end = buffer.as_mut_ptr().add(buffer_width);
+            CFrameItem::new_write(bit_width, buffer_end)
         }
-        let mut src_buf = vec![0 as UWORD; a_frame_size + b_frame_size];
-        let src_ptr_end = unsafe { src_buf.as_mut_ptr().add(a_frame_size) }; // A frame write
-        let src_ptr = src_buf.as_mut_ptr(); // A read frame at ptr start
-        let dst_ptr_begin = unsafe { src_buf.as_mut_ptr().add(a_frame_size) }; // B read frame at ptr begin
-        let dst_ptr_end = unsafe { src_buf.as_mut_ptr().add(a_frame_size + b_frame_size) }; // B write frame at ptr end
-
-        // For jet from type A -> B
-        // Jets execution: There is single buffer with a_frame_size + b_frame_size UWORDs
-        // ------[ A read frame     ][    B write frame  ]---
-        //       ^ src_ptr         ^src_ptr_end(dst_ptr_begin)      ^ dst_ptr_end
-        // 1. Write into C bitmachine using A write frame(= src_ptr_end)
-        // Precondition satisfied: src_ptr_end is one past the end of slice of UWORDs for A.
-        let mut a_frame = unsafe { CFrameItem::new_write(src_ty_bit_width, src_ptr_end) };
-        for _ in 0..src_ty_bit_width {
-            let bit = self.read_bit();
-            unsafe {
-                c_writeBit(&mut a_frame, bit);
+
+        /// Return the value of the given type that is contained in the buffer.
+        fn read_output_frame(mac: &mut BitMachine, bit_width: usize, buffer: &[UWORD]) {
```

Comment:
In 494a232fa8e6144be845940a1bfe22c88e98359e:

Could this simply be a method on `Frame`? The frame itself should know its width now.

As written, is this unsafe if the user passes too large a value for `bit_width`?

Reply:
`Frame`s don't own their data so we would have to pass the data as a parameter, which is messy.